src/main/java/org/candlepin/util/Util.java

```diff
@@ -266,30 +269,18 @@ public static String readFile(InputStream is) {
     }
 
     public static String hash(String password) {
-        //This is secure because even if the salt is known, a cracker
-        //would still need to generate their own rainbow table, which
-        //is the same as brute-forcing the password in the first place.
-
-        String salt = "b669e3274a43f20769d3dedf03e9ac180e160f92";
-        String combined = salt + password;
-
-        MessageDigest md;
-        try {
-            md = MessageDigest.getInstance("SHA-1");
-        }
-        catch (NoSuchAlgorithmException nsae) {
-            throw new RuntimeException(nsae);
-        }
-
-        try {
-            md.update(combined.getBytes("UTF-8"), 0, combined.length());
-        }
-        catch (UnsupportedEncodingException uee) {
-            throw new RuntimeException(uee);
-        }
-
-        byte[] sha1hash = md.digest();
-        return new String(Hex.encodeHex(sha1hash));
+        byte[] salt = generateSalt();
```

Comment:
But this is what the salt is supposed to be doing: each user's password needs to be generated with a unique hash. Two different users using the same plaintext password should result in 2 different hashes exactly because of the salt, and thats a security feature.

And I was also confused by the way this is implemented at first, because I expected the salt unique to that user to be looked up from the cp_user table in the db, along with their hashed password, but apparently this is how bcrypt is supposed to work. You don't compare the hashes by re-using the Util.hash() function on the plaintext. You only use Util.hash() to do the initial hashing, then you use `OpenBSDBCrypt.checkPassword` to check the password. Bcrypt is storing the salt along with the hash itself in the db, so when `checkPassword` is called, it is internally calling `OpenBSDBCrypt.generate` along with the original salt for that user, and checks if its the same.

Reply:
That sounds like not a bad idea to create UtilTest.java.

For a better understanding of how bcrypt works. We will see how the hash looks like:
```
$2a$12$R9h/cIPz0gi.URNNX3kh2OPST9/PgBkqquzi.Ss7KIUgO2t0jWMUW
 \__/\__/\___________________________/\_______________________________________/
 Alg Cost            Salt                                                Hash
``` 
- Alg: Specifying hash algorithm that is being used (We are using crypt_blowfish algorithm. Fixed one, you can see it that is 2y. crypt_blowfish had a bug, but in version 2x.)
- Cost: Input cost. The number is saying how hard it will be to brute force the password. Right now there is recommended value between 10 and 12, but we can scale it up in the future up to 31.
- Salt: base-64 encoding of the salt that was used to hash the password
- Hash: base-64 encoding of hashed password 

So that why is your example failing due to how bcrypt work. We can not use standard equals anymore. We need to compare hashes how Nikos describe in comment above.

Reply:
I think the issue may be that a method called `Util.hash()` could be used for other purposes than generating a password, so maybe we should rename it as `hashPassword` (since it is not actually used for anything other than a password right now)

Reply:
Done (renamed and add tests)

Reply:
Yes, I'm aware of how salts work, but you still need to be able to consistently generate the same output for the same input when you're using a hash function. In this case, our salt is generated randomly on every invocation, with no way to ever specify the salt to get back to the same output.

Moreover, relying on the specific of bcrypt's output string to expose the salt (which is kinda weird, but whatever... it works) to regenerate it later is also problematic. We're now tying a generic function to a specific one without documentation or any other mechanism to prevent more maintenance like this in the future.

Reply:
I do not think that static salt is a good idea. in my opinion, it is desirable to have a different hash for two identical passwords as long as we have also the salt stored in the database.

Reply:
At no point here have I said or implied the use of a static salt. 

If we're going to have a hashing function that applies a salt, and that function is generic, the salt should be an input to the method along with our desired hash target. At that point you can then verify the function's behavior by providing the same inputs and getting the same output every time, even with a randomly generated salt.

Reply:
Extracting the salt is more difficult than one would expect, because they encode the salt using non-standard base64 and so we would have to copy their entire conversion tables. so I checked the consistency by generating the salt in a test and then using it to hash one string. Or we can extract the old code and mark it as not to be used in production, like Crog mentioned in comment. What sounds like a better idea to you? @nikosmoum @Ceiu

Reply:
If bcrypt is hard to work with, we don't have to use it. We could just use SHA-512 (in the same way we used SHA-1 before), with generating salt dynamically and persisting it in a new cp_user column. @Ceiu what do you think?

Reply:
if we don't want to extract the salt and use their method  `OpenBSDBCrypt.checkPassword` for comparison, it should be fine. 

Theoretically we can extract the salt but it would mean that we would have to copy a lot of code from the `OpenBSDBCrypt.java` and then maintain it.

Reply:
If the problem here is that `Util.hash` is supposed to be a generic hashing function, and we're repurposing it, then I'd say just rename it to `Util.encryptPassword`. From what I see this hash function was not used for anything other than generating and checking the validity of passwords, so we don't need one such generic hashing function

Reply:
adding javadoc to this method to explain how it works and how it should be used alongside `OpenBSDBCrypt.checkPassword` and we should be fine?